src/vm/costs/mod.rs

```diff
@@ -93,37 +112,95 @@ impl CostTracker for () {
     fn reset_memory(&mut self) {}
 }
 
+#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
+pub struct ClarityCostFunctionReference {
+    pub contract_id: QualifiedContractIdentifier,
+    pub function_name: String,
+}
+
+impl ClarityCostFunctionReference {
+    fn new(id: QualifiedContractIdentifier, name: String) -> ClarityCostFunctionReference {
+        ClarityCostFunctionReference { contract_id: id, function_name: name }
+    }
+}
+
+type ClarityCostContract = &'static str;
+
 #[derive(Debug, Clone, PartialEq)]
 pub struct LimitedCostTracker {
+    cost_function_references: HashMap<&'static ClarityCostFunction, ClarityCostFunctionReference>,
+    cost_contracts: HashMap<QualifiedContractIdentifier, ContractAST>,
     total: ExecutionCost,
     limit: ExecutionCost,
     memory: u64,
     memory_limit: u64,
+    free: bool,
 }
 
 #[derive(Debug, PartialEq, Eq)]
 pub enum CostErrors {
+    CostComputationFailed(String),
     CostOverflow,
     CostBalanceExceeded(ExecutionCost, ExecutionCost),
     MemoryBalanceExceeded(u64, u64),
 }
 
 impl LimitedCostTracker {
+    //TODO: require cost functions/contracts to be included
```

Comment:
If possible, can you address this before merging?  We want to make invalid states unrepresentable.

Reply:
Some context.  We already know that the cost tracker's cost functions are going to come from a set of voted-upon public methods in various contracts, which will be tracked on a per-chaintip basis.  This means that the Clarity VM will already have all the knowledge it needs to load up and instantiate the cost functions when it begins processing the next block -- it will already need to look them up from the current chain tip, for example.  It's entirely reasonable to me to simply have this method take the set of cost functions as an argument.

Reply:
I think that that is true, but I don't think it should be implemented in this PR -- implementation of the chain state passing in cost functions is planned for a separate PR.